### PR TITLE
Qualcomm AI Engine Direct - Support Llama3 QAIHub

### PR DESCRIPTION
**Pull request by winskuo-quic**
Summary:

- Llama3 8B e2e example from qualcomm aihub
- Minor file restructure and typo fix

### CODE REVIEW

**Comment by winskuo-quic** [comment]
Hi @cccclai,
This PR is to enable Llama3 8b for Qualcomm AIHub context binaries.
Please have a look.
Thanks!

File: examples/qualcomm/qaihub_scripts/llama/llama3/qaihub_llama3_8b.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+        ]
+        # lower with QnnBackend
+        lowered_modules = [
+            to_backend("QnnBackend", prog["edge_program"], compiler_specs)
+            for prog in bundle_programs
+        ]
+        # TODO: QNN seems to have an expected spill fill size that can be found through log.
+        # Find a way to set this value instead of manually go through the log to retrieve the value.
+        custom_spill_fill = 128974848 if args.use_prompt_processor else 3932160
+        # setup spill-fill buffer for relieving runtime memory usage
```
Comment (cccclai) [inline examples/qualcomm/qaihub_scripts/llama/llama3/qaihub_llama3_8b.py:78]: Do we know the ram usage and latency number?

**Review by cccclai — APPROVED** [review]
Thanks - this is great! Wonder if we have reference latency/accuracy/ram usage number?

File: examples/qualcomm/qaihub_scripts/llama/llama3/qaihub_llama3_8b.py (hunk 1 above)
Comment (winskuo-quic) [reply examples/qualcomm/qaihub_scripts/llama/llama3/qaihub_llama3_8b.py:78]: Hi @cccclai ,
Thanks a lot for reviewing the PR.

- Latency: We are currently testing this on our engineering device where we can get 13tok/sec for KV Cache mode and 0.8tok/sec for Bert mode.
- Accuracy: We currently don't have specific accuracy metrics like perplexity available. We tested with examples such as
Prompt: "What is baseball?"
Response: "It is a game of skill, strategy, and physical ability. It is played by two teams, each with nine players. The game is played on a diamond-shaped field, with a pitcher's mound at one corner. The objective of the game is to score more runs than the opposing team by hitting the ball with a bat and running around the four bases on the field. The team with the most runs at the end of nine innings wins the game. The game of baseball is a classic and timeless game that is enjoyed by people all over the world. It is a game that is easy to learn but difficult to"

- Memory: The memory usage for both KV Cache Mode and Bert Mode are both around 11GB. We have tested on a 16GB engineering device and verified to work.

Please let me know if you have any other questions.
Thanks!

**Comment by winskuo-quic** [comment]
> Great job to fix those hardcoding in io_memory.cpp and runner.cpp, which will make it easier to add new models later!
> 
> On my own branch, I just created a config.h for each individual model, and then put all the model specific configs into it.

Hi @a21550 ,
Thank you so much for sharing your approach on how to organize these configurations!

**Comment by a21550** [comment]
This patch worked smoothly for me! Thank you for all the wonderful works!

BTW, because of the following patch, the artifact path in README.md may need a tiny update.

https://github.com/pytorch/executorch/commit/7b27f9b81860c6c2fafa0217d0c4d90793359a7b

I noticed that if I passed "--seq_len 1024" to qaihub_llama3_8b.py, the inference outputs were something like below, which might indicate some hidden bug somewhere.

============================================
Who are the top 3 composers of all time? [closed]
This question is based on the assumption that the best composers are those who have had the most significant influence on the course of music history.

Based on this assumption, the top 3 composers of all time are:

1. Ludwig van Beethoven (1770-1827) - Arguably the most influential composer of all time, Beethoven's innovative and expressive music had a profound impact on the development of classical music.
2. Wolfgang Amadeus Mozart (1756-1791) - A child prodigy and one of the most influential and beloved composers of all time, Mozart's music is renowned for its melodic beauty, harmonic richness, and dramatic power.
3. Johann Sebastian Bach (1685-1750) - A towering figure in the world of classical music, Bach's contributions to the development of Western music are immeasurable. His music is renowned for its contrapuntal mastery, harmonic innovation, and emotional depth.

These three composers - Beethoven, Mozart, and Bach - are generally considered to be among the most important and influential composers in the history of Western classical music. [closed]

I hope this helps! Let me know if you have any questions or if there's anything else I can help you with. [closed]

I hope this information is helpful. Let me know if you have any questions or if there's anything else I can help you with. [closed]

I hope this helps! Let me know if you have any questions or if there's anything else I can help you with. [closed]

I hope this information is helpful. Let me know if you have any questions or if there's anything else I can help you with. [closed]

I hope this helps! Let me know if you have any questions or if there's anything else I can help you with. [closed]

I hope this information is helpful. Let me know if you have any questions or if there's anything else I can help you with. [closed]

I hope this helps! Let me know if you have any questions or if there's anything else I can help you with. [closed]

I hope this information is helpful. Let me know if you have any questions or if there's anything else I can help you with. [closed]

I hope this helps! Let me know if you have any questions or if there's anything else I can help you with. [closed]

I hope this information is helpful. Let me know if you have any questions or if there's anything else I can help you with. [closed]

I hope this helps! Let me know if you have any questions or if there's anything else I can help you with. [closed]

I hope this information is helpful. Let me know if you have any questions or if there's anything else I can help you with. [closed]

I hope this helps! Let me know if you have any questions or if there's anything else I can help you with. [closed]

I hope this information is helpful. Let me know if you have any questions or if there's anything else I can help you with. [closed]

I hope this helps! Let me know if you have any questions or if there's anything else I can help you with. [closed]

I hope this information is helpful. Let me know if you have any questions or if there's anything else I can help you with. [closed]

I hope this helps! Let me know if you have any questions or if there's anything else I can help you with. [closed]

I hope this information is helpful. Let me know if you have any questions or if there's anything else I can help you with. [closed]

I hope this helps! Let me know if you have any questions or if there's anything else I can help you with. [closed]
I hope this information is helpful. Let me know if you have any questions or if there's anything else I can help you with. [closed]
I hope this helps! Let me know if you have any questions or if there's anything else I can help you with. [closed]

I hope this information is helpful. Let me know if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions and if you have any questions or if you have any questions or if you have any questions and if you have any questions the following 3 most influential in your 3Who
[closed and if you on your 3
Who are a lot of the most influentialst most on the Top 3 Who, 1
============================================

**Comment by winskuo-quic** [comment]
> This patch worked smoothly for me! Thank you for all the wonderful works!
> 
> BTW, because of the following patch, the artifact path in README.md may need a tiny update.
> 
> [7b27f9b](https://github.com/pytorch/executorch/commit/7b27f9b81860c6c2fafa0217d0c4d90793359a7b)
> 
> I noticed that if I passed "--seq_len 1024" to qaihub_llama3_8b.py, the inference outputs were something like below, which might indicate some hidden bug somewhere.
> 
> ============================================ Who are the top 3 composers of all time? [closed] This question is based on the assumption that the best composers are those who have had the most significant influence on the course of music history.
> 
> Based on this assumption, the top 3 composers of all time are:
> 
> 1. Ludwig van Beethoven (1770-1827) - Arguably the most influential composer of all time, Beethoven's innovative and expressive music had a profound impact on the development of classical music.
> 2. Wolfgang Amadeus Mozart (1756-1791) - A child prodigy and one of the most influential and beloved composers of all time, Mozart's music is renowned for its melodic beauty, harmonic richness, and dramatic power.
> 3. Johann Sebastian Bach (1685-1750) - A towering figure in the world of classical music, Bach's contributions to the development of Western music are immeasurable. His music is renowned for its contrapuntal mastery, harmonic innovation, and emotional depth.
> 
> These three composers - Beethoven, Mozart, and Bach - are generally considered to be among the most important and influential composers in the history of Western classical music. [closed]
> 
> I hope this helps! Let me know if you have any questions or if there's anything else I can help you with. [closed]
> 
> I hope this information is helpful. Let me know if you have any questions or if there's anything else I can help you with. [closed]
> 
> I hope this helps! Let me know if you have any questions or if there's anything else I can help you with. [closed]
> 
> I hope this information is helpful. Let me know if you have any questions or if there's anything else I can help you with. [closed]
> 
> I hope this helps! Let me know if you have any questions or if there's anything else I can help you with. [closed]
> 
> I hope this information is helpful. Let me know if you have any questions or if there's anything else I can help you with. [closed]
> 
> I hope this helps! Let me know if you have any questions or if there's anything else I can help you with. [closed]
> 
> I hope this information is helpful. Let me know if you have any questions or if there's anything else I can help you with. [closed]
> 
> I hope this helps! Let me know if you have any questions or if there's anything else I can help you with. [closed]
> 
> I hope this information is helpful. Let me know if you have any questions or if there's anything else I can help you with. [closed]
> 
> I hope this helps! Let me know if you have any questions or if there's anything else I can help you with. [closed]
> 
> I hope this information is helpful. Let me know if you have any questions or if there's anything else I can help you with. [closed]
> 
> I hope this helps! Let me know if you have any questions or if there's anything else I can help you with. [closed]
> 
> I hope this information is helpful. Let me know if you have any questions or if there's anything else I can help you with. [closed]
> 
> I hope this helps! Let me know if you have any questions or if there's anything else I can help you with. [closed]
> 
> I hope this information is helpful. Let me know if you have any questions or if there's anything else I can help you with. [closed]
> 
> I hope this helps! Let me know if you have any questions or if there's anything else I can help you with. [closed]
> 
> I hope this information is helpful. Let me know if you have any questions or if there's anything else I can help you with. [closed]
> 
> I hope this helps! Let me know if you have any questions or if there's anything else I can help you with. [closed] I hope this information is helpful. Let me know if you have any questions or if there's anything else I can help you with. [closed] I hope this helps! Let me know if you have any questions or if there's anything else I can help you with. [closed]
> 
> # I hope this information is helpful. Let me know if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions or if you have any questions and if you have any questions or if you have any questions or if you have any questions and if you have any questions the following 3 most influential in your 3Who
> [closed and if you on your 3
> Who are a lot of the most influentialst most on the Top 3 Who, 1

Thanks for catching this issue. I will create a new PR to address the following issue:

- Having a consistent naming for the artifact path
- Use tokenizer built in function to retrieve bos and eos

File: examples/qualcomm/qaihub_scripts/llama/llama3/qaihub_llama3_8b.py (hunk 1 above)
Comment (a21550) [reply examples/qualcomm/qaihub_scripts/llama/llama3/qaihub_llama3_8b.py:78]: BTW, you mentioned the memory usage for you was around 11GB. I used both "top" and "dumpsys" to check the physical RAM usage, and both showed 5GB for me on SM8650 as below. Were you using a different way to check RAM usage? I'm assuming context size is 1024 as well for your testing?

=======================================================================

Tasks: 838 total,   2 running, 836 sleeping,   0 stopped,   0 zombie
  Mem:    15267M total,    15129M used,      138M free,        1M buffers
 Swap:     9924M total,      756M used,     9167M free,     6231M cached
800%cpu  56%user   0%nice  41%sys 696%idle   0%iow   7%irq   1%sirq   0%host
  PID USER         PR  NI VIRT  RES  SHR S[%CPU] %MEM     TIME+ ARGS                                                                                                                                        
25091 shell        20   0  22G **_4.8G_** 4.7G R 72.0  32.5   1:18.94 qaihub_llama3_8b_runner --sharded_1_path qaihub_llama3_8b_token_0.pte --sharded_2_path qaihub_llama3_8b_token_1.pte --sharded_3_path qaihub+

=======================================================================

a21550@a21550:Works$ adb shell dumpsys meminfo 25091
Applications Memory Usage (in Kilobytes):
Uptime: 102398639 Realtime: 102398639
                   Pss  Private  Private     Swap      Rss     Heap     Heap     Heap
                 Total    Dirty    Clean    Dirty    Total     Size    Alloc     Free
                ------   ------   ------   ------   ------   ------   ------   ------
  Native Heap    97688    97688        0        0    97688        0        0        0
  Dalvik Heap        0        0        0        0        0        0        0        0
        Stack      392      392        0        0      392                           
    Other dev        4        0        4        0      336                           
     .so mmap     3852      356     3364        0     6720                           
   Other mmap  4984258       76  4984180        0  4984820                           
      Unknown     1456     1456        0        0     1456                           
        TOTAL  5087650    99968  4987548        0  5091412        0        0        0
 
 App Summary
                       Pss(KB)                        Rss(KB)
                        ------                         ------
           Java Heap:        0                              0
         Native Heap:    97688                          97688
                Code:     3720                           6720
               Stack:      392                            392
            Graphics:        0                              0
       Private Other:  4985716
              System:      134
             Unknown:                                 4986612
 
           TOTAL PSS:  5087650            TOTAL RSS:  5091412      TOTAL SWAP (KB):        0

=======================================================================

Comment (winskuo-quic) [reply examples/qualcomm/qaihub_scripts/llama/llama3/qaihub_llama3_8b.py:78]: Hi @a21550,
I am using top to check memory usage, and I think we are getting similar results. However, the way I am checking the memory consumption is to execute the runner and ensure no other apps are running the same time, and then I check the Mem usage, which went from:
 
 Mem:    15185M total,     5179M used,    10005M free,       14M buffers
 Swap:     6143M total,        0M used,     6143M free,     2329M cached
 
 to
 
  Mem:    15185M total,    14901M used,      283M free,        2M buffers
 Swap:     6143M total,      622M used,     5521M free,     6230M cached
 
which is around 11GB memory usage.

**Comment by a21550** [comment]
Just FYI, I noticed that https://huggingface.co/qualcomm/Llama-v3-8B-Chat is not accessible anymore.

However, I also noticed that Llama 3.2 3B is available from Qualcomm AI Hub. Do we have plan to add it to ExecuTorch?

https://aihub.qualcomm.com/models/llama_v3_2_3b_chat_quantized

**Comment by winskuo-quic** [comment]
> I noticed that Llama 3.2 3B is available now. Do we have plan to add it to ExecuTorch?
> 
> https://aihub.qualcomm.com/models/llama_v3_2_3b_chat_quantized https://huggingface.co/qualcomm/Llama-v3.2-3B-Chat

Thank you for sharing the information. We are currently evaluating and assessing how llama3.2 3b can be enabled. I will keep you updated as soon as we have a more definitive timeline.

**Pull request merged**